services/scheduler/src/main/java/org/mule/service/scheduler/internal/DefaultScheduler.java

```diff
@@ -0,0 +1,294 @@
+  private Set<BaseSchedulerTaskDecorator> currentTasks = synchronizedSet(new HashSet<>());
+
+  private volatile boolean shutdown = false;
+
+  /**
+   * Wait condition to support awaitTermination
+   */
+  private final CountDownLatch terminationLatch = new CountDownLatch(1);
+
+  /**
+   * @param executor the actual executor that will run the dispatched tasks.
+   * @param scheduledExecutor the executor that will handle the delayed/periodic tasks. This will not execute the actual tasks,
+   *        but will dispatch it to the {@code executor} at the appropriate time.
+   */
+  public DefaultScheduler(ExecutorService executor, ScheduledExecutorService scheduledExecutor) {
+    this.executor = executor;
+    this.scheduledExecutor = scheduledExecutor;
+  }
+
+  @Override
+  public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+    checkShutdown();
+    if (command == null) {
+      throw new NullPointerException();
+    }
+
+    final DefaultSchedulerTaskDecorator decorated = decorateRunnable(command);
+
+    return scheduledExecutor.schedule(() -> executor.execute(decorated), delay, unit);
+  }
+
+  @Override
+  public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
+    checkShutdown();
+    if (callable == null) {
+      throw new NullPointerException();
+    }
+
+    final DefaultSchedulerTaskDecorator decorated = new DefaultSchedulerTaskDecorator(new FutureTask<>(callable), this);
+    currentTasks.add(decorated);
+
+    return (ScheduledFuture<V>) scheduledExecutor.schedule(() -> executor.execute(decorated), delay, unit);
+  }
+
+  @Override
+  public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit) {
+    checkShutdown();
+    if (command == null) {
+      throw new NullPointerException();
+    }
+
+    final DefaultSchedulerTaskDecorator decorated = decorateRunnable(command);
+
+    return scheduledExecutor.scheduleAtFixedRate(() -> executor.execute(decorated), initialDelay, period, unit);
+  }
+
+  @Override
+  public ScheduledFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay, TimeUnit unit) {
+    checkShutdown();
+    if (command == null) {
+      throw new NullPointerException();
+    }
+
+    final DefaultSchedulerTaskDecorator decorated = decorateRunnable(command);
+
+    return scheduledExecutor.scheduleWithFixedDelay(() -> executor.execute(decorated), initialDelay, delay, unit);
+  }
+
+  @Override
+  public void shutdown() {
+    this.shutdown = true;
+    tryTerminate();
+  }
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    shutdown();
+
+    List<Runnable> tasks;
+    try {
+      synchronized (currentTasks) {
+        tasks = new ArrayList<>(currentTasks.size());
+        for (BaseSchedulerTaskDecorator task : currentTasks) {
+          task.stop();
+          tasks.add(task.getDecoratedRunnable());
+        }
+        currentTasks.clear();
+
+        return tasks;
```

Comment:
It seems this implementation will return both pending tasks and those that were executing when shutdown is called or those that are interupted, when it should only return pending tasks.  Task.stop() should probably remove tasks from currentTasks to achieve this but there is potentially a deadlock doing that, that needs to be avoided.